urlshort/handler_test.go

```diff
@@ -0,0 +1,215 @@
+
+	wantErr     bool
+	expectedErr error
+}
+
+type expectedResponse struct {
+	Status int
+	URL    string
+}
+
+func TestParseYAML(t *testing.T) {
+
+	tests := []parseYAMLtests{
+
+		{
+
+			name:  "Check for valid unmarshalling of data",
+			input: []byte(yamlData),
+			expected: []URLShort{
+
+				{
+					Path: "/wilken",
+					URL:  "https://wilkenrivera.com/about",
+				},
+				{
+					Path: "/youtube",
+					URL:  "https://www.youtube.com/watch?v=N9ZWy7xCQ8U&t",
+				},
+
+				{
+					Path: "/shreyash",
+					URL:  "https://www.linkedin.com/in/shreyash-bukkawar-816bb1116/",
+				},
+			},
+
+			wantErr: false,
+		},
+
+		{
+			name: "Send unformatted Yaml and get error",
+
+			input: []byte(`
+			
+			- path: /wilken
+ 			 url: https://wilkenrivera.com/about
+			-path: /youtube
+  			url: https://www.youtube.com/watch?v=N9ZWy7xCQ8U&t
+         	- path: /shreyash
+  			  url: https://www.linkedin.com/in/shreyash-bukkawar-816bb1116/
+			`),
+
+			expected: []URLShort{},
+
+			wantErr:     true,
+			expectedErr: errors.New("found character that cannot start any token"),
+		},
+	}
+
+	for _, currentTest := range tests {
+
+		t.Run(currentTest.name, func(t *testing.T) {
+
+			actualOutput, err := ParseYAML(currentTest.input)
+
+			if err != nil {
+				if currentTest.wantErr {
+
+					if strings.Contains(err.Error(), currentTest.expectedErr.Error()) {
+						//Tests Passed if it conatins error message
+
+						return
+					}
+
+					t.Errorf("Expected error = %v got = %v ", currentTest.expectedErr.Error(), err.Error())
+				}
+				return
+			}
+
+			if !reflect.DeepEqual(actualOutput, currentTest.expected) {
+
+				t.Errorf("Expected = %v got = %v ", currentTest.expected, actualOutput)
+			}
+
+		})
+
+	}
+
+}
+
+func TestYAMLHandler(t *testing.T) {
+
+	tests := []yamlHandlertests{
+
+		{
+			name: "check for a Successful Redirect",
+
+			input: map[string]interface{}{
+
+				"/shreyash": "https://www.linkedin.com/in/shreyash-bukkawar-816bb1116",
+				"/wilken":   "https://wilkenrivera.com/about",
+				"/youtube":  "https://www.youtube.com/watch?v=N9ZWy7xCQ8U&t",
+			},
+			inputPath: "/wilken",
+
+			expected: expectedResponse{
+
+				Status: 302,
+				URL:    "https://wilkenrivera.com/about",
+			},
+
+			wantErr: false,
+		},
+
+		{
+			name: "Give a path that doesnt existin map and get error",
+
+			input: map[string]interface{}{
+
+				"/shreyash": "https://www.linkedin.com/in/shreyash-bukkawar-816bb1116",
+				"/wilken":   "https://wilkenrivera.com/about",
+				"/youtube":  "https://www.youtube.com/watch?v=N9ZWy7xCQ8U&t",
+			},
+			inputPath: "/unknownpath",
+
+			expected: expectedResponse{
+
+				Status: 404,
+				URL:    "",
+			},
+
+			wantErr: false,
+		},
+	}
+
+	for _, currentTest := range tests {
+
+		t.Run(currentTest.name, func(t *testing.T) {
+
+			//url := currentTest.input[currentTest.inputPath].(string)
+			//fmt.Printf("url is %v \n", url)
+			req, err := http.NewRequest("GET", currentTest.inputPath, nil)
```

Comment:
On the right track here. You can simplify this further to use only the  httptest.ResponseRecoder. There is an example on the httptest pkg you can refer to https://pkg.go.dev/net/http/httptest#ResponseRecorder 

```
	req := httptest.NewRequest("GET", "http://example.com/foo", nil)
	w := httptest.NewRecorder()
	handler(w, req)
	resp := w.Result()
	body, _ := ioutil.ReadAll(resp.Body)
	fmt.Println(resp.StatusCode)
	fmt.Println(resp.Header.Get("Content-Type"))
	fmt.Println(string(body))
```